### PR TITLE
Add support for IPv6 

### PR DESCRIPTION
**Pull request by jjngx**
### Proposed changes

Add support for IPv6

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [X] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/main/CONTRIBUTING.md) doc
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have checked that all unit tests pass after adding my changes
- [X] I have updated necessary documentation
- [X] I have rebased my branch onto main
- [X] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### CODE REVIEW

**Review by eufinco — COMMENTED** [review]
No unit tests with IPv6 addresses?

File: build/Dockerfile [hunk 1]
```diff
@@ -50,7 +50,7 @@ RUN --mount=type=secret,id=nginx-repo.crt,dst=/etc/ssl/nginx/nginx-repo.crt,mode
 	&& apt-get install -y zlib1g \
 	&& curl -fsSL https://cs.nginx.com/static/keys/nginx_signing.key | gpg --dearmor > /etc/apt/trusted.gpg.d/nginx_signing.gpg \
 	&& curl -fsSL -o /etc/apt/apt.conf.d/90pkgs-nginx https://cs.nginx.com/static/files/90pkgs-nginx \
-	&& DEBIAN_VERSION=$(awk -F '=' '/^VERSION_CODENAME=/ {print $2}' /etc/os-release) \
+	&& DEBIAN_VERSION=bullseye \
```
Comment (pleshakov) [inline build/Dockerfile:53]: Can we revert this change?

File: build/Dockerfile [hunk 2]
```diff
@@ -59,7 +59,16 @@ RUN --mount=type=secret,id=nginx-repo.crt,dst=/etc/ssl/nginx/nginx-repo.crt,mode
 	&& rm -rf /var/lib/apt/lists/*
 
 
-############################################# Base image for Debian with NGINX Plus and App Protect WAF/DoS #############################################
```
Comment (pleshakov) [inline build/Dockerfile:62]: Do we need this commented out part?

File: internal/configs/version1/nginx-plus.tmpl [hunk 3]
```diff
@@ -197,6 +199,7 @@ http {
     # NGINX Plus APIs
     server {
         listen {{.NginxStatusPort}};
+        listen [::]:{{.NginxStatusPort}};
```
Comment (pleshakov) [inline internal/configs/version1/nginx-plus.tmpl:202]: I think it makes sense to change the default value of https://github.com/nginxinc/kubernetes-ingress/blob/main/cmd/nginx-ingress/main.go#L134 to include the IPV6 localhost.  That value is used in this server in  the allow directive:
```
        {{range $value := .NginxStatusAllowCIDRs}}
        allow {{$value}};{{end}}
        deny all;
        location /api {
            api write=off;
        }
```

currently, the default value is `127.0.0.1`.

this endpoint is used  to access the dashboard https://docs.nginx.com/nginx-ingress-controller/logging-and-monitoring/status-page/#accessing-live-activity-monitoring-dashboard 

same for NGINX OSS.

File: internal/configs/version1/nginx.tmpl [hunk 4]
```diff
@@ -107,13 +107,15 @@ http {
...
         {{if .TLSPassthrough}}
         listen unix:/var/lib/nginx/passthrough-https.sock ssl default_server{{if .HTTP2}} http2{{end}} proxy_protocol;
         set_real_ip_from unix:;
         real_ip_header proxy_protocol;
         {{else}}
         listen 443 ssl default_server{{if .HTTP2}} http2{{end}}{{if .ProxyProtocol}} proxy_protocol{{end}};
+        listen [::]:443 ssl default_server{{if .HTTP2}} http2{{end}}{{if .ProxyProtocol}} proxy_protocol{{end}};
         {{end}}
 
         {{if .SSLRejectHandshake}}
```
Comment (pleshakov) [inline internal/configs/version1/nginx.tmpl:121]: listen directive in ` {{- if .NginxStatus}}` needs to be updated similarly to NGINX Plus.

File: tests/kind/kind-config-template.yaml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+networking:
+  ipFamily: {{.IPFamily}}
+  apiServerAddress: {{.APIServerAddress}}
+  #apiServerPort: 6443
```
Comment (pleshakov) [inline tests/kind/kind-config-template.yaml:6]: do we need this parameter commented out? perhaps better to remove all together?

**Review by pleshakov — CHANGES_REQUESTED** [review]
Hi @jjngx 

Please see my comments and suggestions.

Note that the review doesn't cover the changes added to the branch and covered in this PR https://github.com/nginxinc/kubernetes-ingress/pull/2577 

Could we also double check that the examples that provision NGINX backends or configure NGINX using snippets (`grep -R "listen" examples`) also work in IPv6 mode?

File: build/Dockerfile (hunk 1 above)
Comment (jjngx) [reply build/Dockerfile]: @pleshakov reverted

File: build/Dockerfile (hunk 2 above)
Comment (lucacome) [reply build/Dockerfile:62]: I think there's still something left in line 62

**Review by lucacome — APPROVED** [review]
I think you can revert `build/Dockerfile`

**Review by pleshakov — APPROVED** [review]
👍
